### PR TITLE
fix(Webhook Node): Do not create binary data when there is no data in the request

### PR DESCRIPTION
**Pull request by michael-radency**
https://linear.app/n8n/issue/NODE-980/do-not-create-binary-data-for-webhooks-when-there-is-no-data-in-the

related:
https://github.com/n8n-io/n8n/pull/7804/files#r1422641833

### CODE REVIEW

File: packages/nodes-base/nodes/Webhook/Webhook.node.ts [hunk 1]
```diff
@@ -126,7 +126,12 @@ export class Webhook extends Node {
 		}
 
 		const nodeVersion = context.getNode().typeVersion;
-		if (nodeVersion > 1 && !req.body && !options.rawBody) {
+		if (
+			nodeVersion > 1 &&
+			['POST', 'PUT', 'PATCH'].includes(req.method.toUpperCase()) &&
```
Comment (netroy) [inline packages/nodes-base/nodes/Webhook/Webhook.node.ts:131]: This still has the issue that if I create a POST webhook, and trigger it without any payload, `.binary` is still set, but there is no actual binary data file, and the `workflowExecuteAfter` hook still throws an error.
I'm not familiar enough the new code in that hook, so maybe @ivan could help us figure out of the `restoreBinaryDataId` call is still needed.
If we still need to execute that code, then we should update this code to delete `returnItem.binary` when the binary data is empty.

**Review by netroy — CHANGES_REQUESTED** [review]

File: packages/nodes-base/nodes/Webhook/Webhook.node.ts (hunk 1 above)
Comment (ivov) [reply packages/nodes-base/nodes/Webhook/Webhook.node.ts:131]: `restoreBinaryDataId()` is meant for production POST webhooks that accept binary data when BE is in a binary data mode that persists binary data. This is an edge case where the execution ID is not known at the time of writing a binary data file, so we need to restore the ID after the fact.

Using this branch I set filesystem mode, activated a production POST webhook, sent a POST request with no payload, and I'm not seeing any error. Can you elaborate how to trigger this error?

Comment (michael-radency) [reply packages/nodes-base/nodes/Webhook/Webhook.node.ts:131]: @netroy 

with fix in this PR, binary property created only if binaryData exist
image

added same fix to 'multipart/form-data'
image

Comment (netroy) [reply packages/nodes-base/nodes/Webhook/Webhook.node.ts]: updated the check to remove the http method, and instead check if the generated temporary file has any content instead.

File: packages/nodes-base/nodes/Webhook/Webhook.node.ts [hunk 2]
```diff
@@ -283,20 +286,23 @@ export class Webhook extends Node {
...
+				const fileName = req.contentDisposition?.filename ?? uuid();
+				const binaryData = await context.nodeHelpers.copyBinaryFile(
+					binaryFile.path,
+					fileName,
+					req.contentType ?? 'application/octet-stream',
+				);
+
+				if (!binaryData.data) {
+					return { workflowData: [[returnItem]] };
+				}
```
Comment (netroy) [inline packages/nodes-base/nodes/Webhook/Webhook.node.ts:301]: I'm guessing this check will never be true now.

**Review by netroy — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.21.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.21.0)

**Pull request merged**